File: app/src/main/java/com/remotearthsolutions/expensetracker/adapters/ExpenseListAdapter.java
package com.remotearthsolutions.expensetracker.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.adapters.viewholder.DateSectionedViewHolder;
import com.remotearthsolutions.expensetracker.adapters.viewholder.ExpenseListViewHolder;
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel;
import com.remotearthsolutions.expensetracker.databaseutils.models.DateModel;
import com.remotearthsolutions.expensetracker.databaseutils.models.dtos.CategoryExpense;
import com.remotearthsolutions.expensetracker.utils.SharedPreferenceUtils;

import java.util.List;

public class ExpenseListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int HEADER_ENABLE = 1;
    private static final int HEADER_DISABLE = 2;
    private List<CategoryExpense> categoryExpenseList;
    private String currencySymbol;
    private ExpenseListAdapter.OnItemClickListener listener;

    public ExpenseListAdapter(List<CategoryExpense> categoryExpenseList, String currencySymbol) {
        this.categoryExpenseList = categoryExpenseList;
        this.currencySymbol = currencySymbol;
    }

    @Override
    public int getItemViewType(int position) {
        CategoryExpense item = categoryExpenseList.get(position);
        if (item.isHeader) {
            return HEADER_ENABLE;
        } else {
            return HEADER_DISABLE;
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == HEADER_ENABLE) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_sectioned_date, parent, false);
            DateSectionedViewHolder dateSectionedViewHolder = new DateSectionedViewHolder(view);
            return dateSectionedViewHolder;
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.custom_all_expense, parent, false);
            return new ExpenseListViewHolder(view,listener);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        CategoryExpense item = categoryExpenseList.get(position);

        if (holder instanceof DateSectionedViewHolder) {
            DateSectionedViewHolder viewHolder = (DateSectionedViewHolder) holder;
            viewHolder.bind(new DateModel(item.getCategory_name()));
        } else {
            ExpenseListViewHolder viewHolder = (ExpenseListViewHolder) holder;
            viewHolder.bind(item, currencySymbol);
        }

    }

    @Override
    public int getItemCount() {
        return categoryExpenseList == null ? 0 : categoryExpenseList.size();
    }

    public void setOnItemClickListener(ExpenseListAdapter.OnItemClickListener listener) {
        this.listener = listener;
    }

    public interface OnItemClickListener {
        void onItemClick(CategoryExpense categoryExpense);
    }

}

File: app/src/main/java/com/remotearthsolutions/expensetracker/adapters/ExpenseListAdapter.kt
package com.remotearthsolutions.expensetracker.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.adapters.viewholder.DateSectionedViewHolder
import com.remotearthsolutions.expensetracker.adapters.viewholder.ExpenseListViewHolder
import com.remotearthsolutions.expensetracker.databaseutils.models.DateModel
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryExpense

class ExpenseListAdapter(
    private val categoryExpenseList: List<CategoryExpense>?,
    private val currencySymbol: String
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var listener: OnItemClickListener? =
        null

    override fun getItemViewType(position: Int): Int {
        val item = categoryExpenseList!![position]
        return if (item.isHeader) {
            HEADER_ENABLE
        } else {
            HEADER_DISABLE
        }
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RecyclerView.ViewHolder {
        return if (viewType == HEADER_ENABLE) {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_sectioned_date, parent, false)
            DateSectionedViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.custom_all_expense, parent, false)
            ExpenseListViewHolder(view, listener!!)
        }
    }

    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int
    ) {
        val item = categoryExpenseList!![position]
        if (holder is DateSectionedViewHolder) {
            holder.bind(DateModel(item.categoryName!!))
        } else {
            val viewHolder = holder as ExpenseListViewHolder
            viewHolder.bind(item, currencySymbol)
        }
    }

    override fun getItemCount(): Int {
        return categoryExpenseList?.size ?: 0
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.listener = listener
    }

    interface OnItemClickListener {
        fun onItemClick(categoryExpense: CategoryExpense?)
    }

    companion object {
        private const val HEADER_ENABLE = 1
        private const val HEADER_DISABLE = 2
    }

}